src/Agent.Listener/DistributedTask.Pipelines/Yaml/PipelineParser.cs

```diff
@@ -299,21 +300,21 @@ private void ResolveTemplates(Process process, String defaultRoot, CancellationT
             {
                 ResolveTemplates(process.Phases, defaultRoot, cancellationToken, ref fileCount);
             }
-            else if (process.Jobs != null)
-            {
-                ResolveTemplates(process.Jobs, defaultRoot, cancellationToken, ref fileCount);
-            }
-            else if (process.Variables != null)
+            else
             {
-                ResolveTemplates(process.Variables, defaultRoot, cancellationToken, ref fileCount);
-            }
-            else if (process.Steps != null)
-            {
-                ResolveTemplates(process.Steps, defaultRoot, cancellationToken, ref fileCount);
+                if (process.Variables != null)
+                {
+                    ResolveTemplates(process.Variables, defaultRoot, cancellationToken, ref fileCount);
+                }
+
+                if (process.Steps != null)
+                {
+                    ResolveTemplates(process.Steps, defaultRoot, cancellationToken, ref fileCount);
+                }
             }
         }
 
-        private void ResolveTemplates(List<IPhase> phases, String defaultRoot, CancellationToken cancellationToken, ref Int32 fileCount)
+        private void ResolveTemplates(IList<IPhase> phases, String defaultRoot, CancellationToken cancellationToken, ref Int32 fileCount)
```

Comment:
I went through and cleaned up all of the method signatures and contracts to use IList and IDictionary instead of List and Dictionary.

The rest of the changes in this file is related to flatting the jobs hierarchy (phase-level execution options now, and steps moved up to phase-level).